### ISSUE TITLE
Issue with custom ops

### ISSUE THREAD

**Issue by tlepley-cadence** [onnx/tensorflow-onnx#330]
I'm trying to setup a flow with custom op from TensorFlow to ONNX.

For instance, I created and saved a TensorFlow model with some custom ops (of op type `MyAdd`). I can reload this model and infer it with TensorFlow (I need to register my TensorFlow custom op before loading the model).

I've seen in tf2onnx an option `--custom-ops` but tf2onnx fails when I try to convert my TensorFlow model to ONNX. I'm then not sure to correctly understand the `--custom-ops` option.

Could you please clarify the current support of custom ops with tf2onnx ?

```
localhost:python -m tf2onnx.convert --custom-ops MyAdd  --input test_graph.pb   --inputs x:0,y:0,z:0   --outputs MyAdd_1:0   --output test_graph.onnx

using tensorflow=1.11.0, onnx=1.4.1, opset=7, tfonnx=0.4.0/717721

2019-02-27 17:47:01.505820: I tensorflow/tools/graph_transforms/transform_graph.cc:318] Applying fold_batch_norms
2019-02-27 17:47:01.505990: I tensorflow/tools/graph_transforms/transform_graph.cc:318] Applying fold_old_batch_norms
Traceback (most recent call last):
  File "/usr/lib64/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/tlepley/git/tensorflow-onnx/tf2onnx/convert.py", line 122, in <module>
    main()
  File "/home/tlepley/git/tensorflow-onnx/tf2onnx/convert.py", line 91, in main
    tf.import_graph_def(graph_def, name='')
  File "/home/tlepley/.local/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/home/tlepley/.local/lib/python3.6/site-packages/tensorflow/python/framework/importer.py", line 418, in import_graph_def
    graph._c_graph, serialized, options)  # pylint: disable=protected-access
tensorflow.python.framework.errors_impl.NotFoundError: Op type not registered 'MyAdd' in binary running on localhost.localdomain. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed.
```

**Comment by zhijxu-MS** [comment]
custom-ops here means you have a **custom op at onnx side** and you want to convert general tf ops into onnx instead of custom op at tensorflow side.

**Comment by tlepley-cadence** [comment]
@zhijxu-MS Thanks for the clarification. I'm nevertheless still not sure to fully understand how the use case you mentioned would work with tf2onnx.

My understanding is that there is a possibility to extend ONNX for instance by simply adding to the ONNX model something like : `opset_import { domain = "com.<company>", op_version = 1 }` . With this declaration, I should be able to add non standard nodes in my ONNX model.

So in the use-case you mentioned, for instance, if I want to convert a standard TF op (let's say depthwise_conv2d for example) into a custom ONNX op with tf2onnx, what would be the command line and what would be get in the generated ONNX model ? I guess I shoud inform tf2onnx about 2 things:
- the TF op type name that I want to keep custom
- The domain this op type would belong to in the ONNX world 
But currently, it seems that we can't express the target ONNX domain with tf2onnx options.
Am I missing something ?

By the way, --custom-ops should get the name of nodes, or the op type name ?
If I want to keep all TF `depthwise_conv2d` nodes as custom ONNX nodes, would the following command work ?
```
python -m tf2onnx.convert \
       --custom-ops depthwise_conv2d \
       --input test_graph.pb   \
       --inputs x:0,y:0,z:0   --outputs MyAdd_1:0 \
       --output test_graph.onnx
```

**Comment by zhijxu-MS** [comment]
just double check the code, it looks like you are right.

but the error above is issued at tensorflow side, do you register the custom op before tensorflow restores the model from your pb file?

**Comment by tlepley-cadence** [comment]
In my first example, where indeed TF fails when tf2onnx attempts to load the model, I did not register the custom op in TF because I was simply running tf2onnx from the command line.

An option may be to call tf2onnx from a python script that would register custom ops in TF prior to calling tf2onnx.convert. I'm going to make a try.

**Comment by tlepley-cadence** [comment]
For my use-case (TF custom op), I just did the try registring the custom op in TF before calling tf2onnx, of tf2onnx now correctly loads the TF model (what was expected), it fails later:

Python script (convert2onnx.py):
```
import tensorflow as tf
import tf2onnx.convert

# Register the TF custom layer
tf.load_op_library('./build/libmy_add.so')

tf2onnx.convert.main()
```

Failure:
```
py3 convert2onnx.py     --custom-ops MyAdd,MyAdd_1     --input test_graph.pb     --inputs x:0,y:0,z:0 --outputs MyAdd_1:0     --output test_graph.onnx
using tensorflow=1.11.0, onnx=1.4.1, opset=7, tfonnx=0.4.0/581031
2019-02-28 14:20:39.946142: I tensorflow/tools/graph_transforms/transform_graph.cc:318] Applying fold_batch_norms
2019-02-28 14:20:39.946277: I tensorflow/tools/graph_transforms/transform_graph.cc:318] Applying fold_old_batch_norms
Traceback (most recent call last):
  File "convert2onnx.py", line 7, in <module>
    tf2onnx.convert.main()
  File "/home/tlepley/git/tensorflow-onnx/tf2onnx/convert.py", line 103, in main
    inputs_as_nchw=args.inputs_as_nchw)
  File "/home/tlepley/git/tensorflow-onnx/tf2onnx/tfonnx.py", line 2509, in process_tf_graph
    mapped_op, unmapped_op = tensorflow_onnx_mapping(g, continue_on_error, custom_op_handlers)
  File "/home/tlepley/git/tensorflow-onnx/tf2onnx/tfonnx.py", line 2316, in tensorflow_onnx_mapping
    func, args = map_info
TypeError: 'function' object is not iterable
```

**Comment by zhijxu-MS** [comment]
[pr338](https://github.com/onnx/tensorflow-onnx/pull/338) will fix the error, you could try it again. and feel free to let me know if any problem.

**Comment by tlepley-cadence** [comment]
I just did a rapid test and it seems to work well now.
Thanks @zhijxu-MS for fixing this so fast !

tf2onnx doesn't fail anymore and when I look at the generated model, my custom ops are declared in the `ai.onnx.converters.tensorflow` domain and this domain is well declared:
```
opset_import {
  domain: "ai.onnx.converters.tensorflow"
  version: 1
}
```

I think it would be nice to have in tf2onnx:
- an option for specifying the custom domain. (maybe even a mapping custom op type name -> domain)
- an option to specify the dynamic libs that implement the custom ops in TF. tf2onnx would then be able to register the TF custom ops itself and I would not need to wrap tf2onnx into another python script.

But I guess these items are more improvements to discuss in another issue. I'm closing this issue and I'll create new issues for improvement proposals.
